### PR TITLE
[WCv2] handle `optionalNamespaces`

### PR DESCRIPTION
**Pull request by estrattonbailey**
Fixes APP-513

## What changed (plus any additional context for devs)
This PR:
- adds a few tests anyway
- further improves mocking
- fixes some small things
- adds support for `optionalNamespaces` via the namespace builder util that WC added recently

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -129,12 +129,12 @@
...
+    "@walletconnect/core": "2.7.3",
     "@walletconnect/legacy-utils": "2.0.0",
-    "@walletconnect/react-native-compat": "2.5.1",
-    "@walletconnect/types": "2.5.1",
-    "@walletconnect/utils": "2.5.1",
-    "@walletconnect/web3wallet": "1.5.0",
+    "@walletconnect/react-native-compat": "2.7.3",
+    "@walletconnect/types": "2.7.3",
+    "@walletconnect/utils": "2.7.3",
+    "@walletconnect/web3wallet": "1.7.1",
```
Comment (estrattonbailey) [inline package.json:137]: Had to upgrade to get the new util

File: src/analytics/event.ts [hunk 2]
```diff
@@ -64,6 +64,11 @@ export const event = {
   rewardsPressedSwappedCard: 'rewards.pressed_swapped_card',
   rewardsPressedBridgedCard: 'rewards.pressed_bridged_card',
   rewardsPressedLeaderboardItem: 'rewards.pressed_leaderboard_item',
+
+  wcNewSessionTimeout: 'New WalletConnect session time out',
+  wcNewSessionRejected: 'Rejected new WalletConnect session',
+  wcNewSessionApproved: 'Approved new WalletConnect session',
+  wcShowingSigningRequest: 'Showing Walletconnect signing request',
```
Comment (estrattonbailey) [inline src/analytics/event.ts:71]: Had to migrate these to new analytics to take advantage of the new mocks

File: src/handlers/__tests__/deeplinks.test.ts [hunk 3]
```diff
@@ -169,7 +169,7 @@ test(`handles https:// protocol for tokens, found asset`, async () => {
 });
 
 test(`handles https:// protocol for tokens, no asset found`, async () => {
-  // @ts-ignore just need a truthy value
+  // @ts-ignore just need a falsy value
```
Comment (estrattonbailey) [inline src/handlers/__tests__/deeplinks.test.ts:172]: Fixed a copy pasta typo

File: src/walletConnect/__tests__/__snapshots__/index.test.ts.snap
Comment (estrattonbailey) [inline src/walletConnect/__tests__/__snapshots__/index.test.ts.snap:1]: Snapshots are just test artifacts, you don't need to review them

File: src/walletConnect/__tests__/index.test.ts [hunk 4]
```diff
@@ -18,8 +54,222 @@ jest.mock('@/screens/Portal');
...
+  await pair({ uri });
+
+  jest.advanceTimersByTime(10_000);
+
+  expect(Navigation.handleAction).toHaveBeenCalled();
+});
 
-// import { parseRPCParams } from '@/walletConnect';
-// import { RPCMethod } from '@/walletConnect/types';
+test.skip(`onSessionProposal`, async () => {
```
Comment (estrattonbailey) [inline src/walletConnect/__tests__/index.test.ts:184]: This test is WIP, might come back later

File: src/walletConnect/index.tsx [hunk 5]
```diff
@@ -16,7 +20,7 @@ import { logger, RainbowError } from '@/logger';
 import { WalletconnectApprovalSheetRouteParams } from '@/redux/walletconnect';
 import Navigation, { getActiveRoute } from '@/navigation/Navigation';
 import Routes from '@/navigation/routesNames';
-import { analytics } from '@/analytics';
+import { analyticsV2 as analytics } from '@/analytics';
```
Comment (estrattonbailey) [inline src/walletConnect/index.tsx:23]: Here, migrated to new analytics, which require the new event config

File: src/walletConnect/index.tsx [hunk 6]
```diff
@@ -94,17 +102,15 @@ let syncWeb3WalletClient:
...
-    core: walletConnectCore,
-    metadata: {
-      name: '🌈 Rainbow',
-      description: 'Rainbow makes exploring Ethereum fun and accessible 🌈',
-      url: 'https://rainbow.me',
-      icons: ['https://avatars2.githubusercontent.com/u/48327834?s=200&v=4'],
-    },
-  })
-);
+export const web3WalletClient = Web3Wallet.init({
```
Comment (estrattonbailey) [inline src/walletConnect/index.tsx:105]: Was double resolving this promise, no need for it

File: src/walletConnect/index.tsx [hunk 7]
```diff
@@ -383,10 +393,7 @@ export async function onSessionProposal(
   const supportedChainIds = chainIds.filter(isSupportedChain);
   const unsupportedChainIds = chainIds.filter(id => !isSupportedChain(id));
 
-  if (
-    (unsupportedChainIds.length && !supportedChainIds.length) ||
-    unsupportedNamespaces.length
-  ) {
+  if (unsupportedChainIds.length) {
```
Comment (estrattonbailey) [inline src/walletConnect/index.tsx:396]: This was a mistake in my logic that `optionalNamespaces` and the builder util clarify. A few lines up we reject if any non eip155 namespaces were required. Here, we reject if any unsupported chains are required.

File: src/walletConnect/index.tsx [hunk 8]
```diff
@@ -452,29 +455,29 @@ export async function onSessionProposal(
...
+              chains: caip2ChainIds,
+              methods: [
+                ...SUPPORTED_SIGNING_METHODS,
+                ...SUPPORTED_TRANSACTION_METHODS,
+              ],
+              events: requiredNamespace.events,
+              accounts: caip2ChainIds.map(id => `${id}:${accountAddress}`),
+            },
+          },
+        });
```
Comment (estrattonbailey) [inline src/walletConnect/index.tsx:476]: Nasty highlighting, but this is the new code here

**Review by skylarbarrera — APPROVED** [review]
🌞

**Review by benisgold — APPROVED** [review]

**Pull request merged**